modules/platforms/dotnet/Apache.Ignite.Core/Impl/Cache/Near/NearCache.cs

```diff
@@ -0,0 +1,261 @@
+
+            _map[pair.Key] = new NearCacheEntry<TV>(
+                pair.Value,
+                GetBoxedAffinityTopologyVersion(affinityTopologyVersion),
+                partition);
+        }
+
+        /** <inheritdoc /> */
+        public void Stop()
+        {
+            _stopped = true;
+            Clear();
+        }
+        
+        /** <inheritdoc /> */
+        public void Clear()
+        {
+            _map.Clear();
+        }
+
+        /** <inheritdoc /> */
+        public void SetThreadLocalPair<TKey, TVal>(TKey key, TVal val)
+        {
+            NearCacheManager.ThreadLocalPair.Value = new KeyValuePair<TK, TV>((TK) (object) key, (TV) (object) val);
+        }
+
+        /** <inheritdoc /> */
+        public void ResetThreadLocalPair()
+        {
+            NearCacheManager.ThreadLocalPair.Value = null;
+        }
+
+        /** <inheritdoc /> */
+        public IEnumerable<ICacheEntry<TKey, TVal>> GetEntries<TKey, TVal>()
+        {
+            if (_stopped)
+            {
+                yield break;
+            }
+
+            foreach (var entry in _map)
+            {
+                yield return new CacheEntry<TKey, TVal>((TKey) (object) entry.Key, (TVal) (object) entry.Value.Value);
+            }
+        }
+
+        /// <summary>
+        /// Checks whether specified cache entry is still valid, based on Affinity Topology Version.
+        /// When primary node changes for a key, GridNearCacheEntry stops receiving updates for that key,
+        /// because reader ("subscription") on new primary is not yet established.
+        /// <para />
+        /// This method is similar to GridNearCacheEntry.valid(). 
+        /// </summary>
+        /// <param name="entry">Entry to validate.</param>
+        /// <typeparam name="TVal">Value type.</typeparam>
+        /// <returns>True if entry is valid and can be returned to the user; false otherwise.</returns>
+        private bool IsValid<TVal>(NearCacheEntry<TVal> entry)
+        {
+            // See comments on _affinityTopologyVersionFunc about boxed copy approach. 
+            var currentVerBoxed = _affinityTopologyVersionFunc();
+            var entryVerBoxed = entry.Version;
+            
+            Debug.Assert(currentVerBoxed != null);
+            
+            if (ReferenceEquals(currentVerBoxed, entryVerBoxed))
+            {
+                // Happy path: true on stable topology.
+                return true;
+            }
+
+            if (entryVerBoxed == null)
+            {
+                return false;
+            }
+
+            var entryVer = (AffinityTopologyVersion) entryVerBoxed;
+            var currentVer = (AffinityTopologyVersion) currentVerBoxed;
+
+            if (entryVer >= currentVer)
+            {
+                return true;
+            }
+
+            var part = entry.Partition;
+            var valid = _affinity.IsAssignmentValid(entryVer, part);
+
+            // Update version or mark as invalid (null).
```

Comment:
Sry for possible question duplicate, can't we set a callback for this affinity changes directly from the Java side?

Reply:
Unfortunately, we can't. Because nothing is invoked on Java side in this case. What happens is:
* Affinity topology version changes (node enters or leaves, etc)
* `GridNearCacheEntry` becomes effectively invalid, because updates are no longer received (primary node changed for the key).
* Data becomes stale, but nothing is being done until user requests this entry
* When user calls cache.Get, `GridNearCacheEntry.valid()` is called, invalid entry is detected, etc

So we have to have the same `valid()` check on .NET side to avoid calling Java on every `Get`